File: app/src/main/java/com/u1fukui/bbs/ui/detail/ThreadDetailActivity.java
package com.u1fukui.bbs.ui.detail;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.databinding.ObservableList;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.view.ViewGroup;

import com.u1fukui.bbs.R;
import com.u1fukui.bbs.customview.BindingHolder;
import com.u1fukui.bbs.customview.ObservableListRecyclerAdapter;
import com.u1fukui.bbs.databinding.ActivityThreadDetailBinding;
import com.u1fukui.bbs.databinding.ViewCommentCellBinding;
import com.u1fukui.bbs.model.BbsThread;
import com.u1fukui.bbs.repository.ThreadRepository;
import com.u1fukui.bbs.ui.BaseActivity;

import javax.inject.Inject;

public class ThreadDetailActivity extends BaseActivity {

    public static final String TAG = ThreadDetailActivity.class.getSimpleName();

    private static final String EXTRA_THREAD = "extra.thread";

    @Inject
    ThreadDetailNavigator navigator;

    @Inject
    ThreadRepository repository;

    private ActivityThreadDetailBinding binding;

    private Adapter adapter;

    private ThreadDetailViewModel viewModel;

    public static Intent createIntent(Context context, BbsThread thread) {
        Intent intent = new Intent(context, ThreadDetailActivity.class);
        intent.putExtra(EXTRA_THREAD, thread);
        return intent;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_thread_detail);

        BbsThread thread = (BbsThread) getIntent().getSerializableExtra(EXTRA_THREAD);
        viewModel = new ThreadDetailViewModel(thread, repository, navigator);
        binding.setViewModel(viewModel);
        initToolbar(binding.toolbar, true);

        initViews();
        viewModel.start();
    }

    private void initViews() {
        adapter = new Adapter(viewModel.getCommentViewModelList());
        binding.recyclerView.setAdapter(adapter);
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(this));
        binding.recyclerView.addItemDecoration(new DividerItemDecoration(this, LinearLayoutManager.VERTICAL));
    }

    @Override
    protected void onDestroy() {
        viewModel.destroy();
        binding.unbind();
        super.onDestroy();
    }

    private static class Adapter extends ObservableListRecyclerAdapter<CommentViewModel, BindingHolder<ViewCommentCellBinding>> {

        public Adapter(@NonNull ObservableList<CommentViewModel> list) {
            super(list);
        }

        @Override
        public BindingHolder<ViewCommentCellBinding> onCreateViewHolder(ViewGroup parent, int viewType) {
            return new BindingHolder<>(parent.getContext(), parent, R.layout.view_comment_cell);
        }

        @Override
        public void onBindViewHolder(BindingHolder<ViewCommentCellBinding> holder, int position) {
            CommentViewModel viewModel = getItem(position);
            ViewCommentCellBinding itemBinding = holder.binding;
            itemBinding.setViewModel(viewModel);
            itemBinding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/u1fukui/bbs/ui/detail/ThreadDetailActivity.kt
package com.u1fukui.bbs.ui.detail

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.u1fukui.bbs.R
import com.u1fukui.bbs.api.ThreadApi
import com.u1fukui.bbs.databinding.ActivityThreadDetailBinding
import com.u1fukui.bbs.model.BbsThread
import com.u1fukui.bbs.paging.comment.CommentPagedListAdapter
import com.u1fukui.bbs.repository.ThreadRepository
import com.u1fukui.bbs.ui.BaseActivity
import javax.inject.Inject

class ThreadDetailActivity : BaseActivity() {

    @Inject
    lateinit var navigator: ThreadDetailNavigator

    @Inject
    lateinit var threadApi: ThreadApi

    private val adapter = CommentPagedListAdapter()

    private val repository: ThreadRepository by lazy {
        ThreadRepository(threadApi)
    }

    private val binding by lazy {
        DataBindingUtil.setContentView<ActivityThreadDetailBinding>(
            this,
            R.layout.activity_thread_detail
        )
    }

    private val bbsThread by lazy {
        intent.getSerializableExtra(EXTRA_THREAD) as BbsThread
    }

    private val viewModel: ThreadDetailViewModel by lazy {
        ViewModelProviders
            .of(
                this, ThreadDetailViewModel.Factory(
                    this,
                    bbsThread,
                    repository,
                    navigator
                )
            )
            .get(ThreadDetailViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        initToolbar(binding.toolbar, true)
        initViews()

        viewModel.commentBindingModelList.observe(this, Observer {
            adapter.submitList(it)
        })
        binding.viewModel = viewModel
    }

    private fun initViews() {
        binding.recyclerView.let {
            it.adapter = adapter
            it.layoutManager = LinearLayoutManager(this)
            it.addItemDecoration(
                DividerItemDecoration(
                    this,
                    LinearLayoutManager.VERTICAL
                )
            )
        }
    }

    override fun onDestroy() {
        binding.unbind()
        super.onDestroy()
    }

    companion object {

        private const val EXTRA_THREAD = "extra.thread"

        @JvmStatic
        fun createIntent(context: Context, thread: BbsThread) =
            Intent(context, ThreadDetailActivity::class.java).apply {
                putExtra(EXTRA_THREAD, thread)
            }
    }
}
